Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,207 @@
+    : IPC::ClientConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    GUI::Application::the()->quit();
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileResponse ClientConnection::request_file(String const& sender, String const& path, int requested_access)
+{
+    if (!path.starts_with("/"sv)) {
+        // TODO: Make developers give us absolute path
+        // or, take the base path as an argument
+        return { -1, Optional<IPC::File> {} };
+    }
+
+    StringBuilder builder;
+    if (requested_access & FileAccess::Read)
+        builder.append('r');
+
+    if (requested_access & FileAccess::Write)
+        builder.append('w');
+
+    if (requested_access & FileAccess::CreateOrRemove)
+        builder.append('c');
+
+    auto access_string = builder.to_string();
+
+    auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} \"{}\" access to \"{}\"?", sender, access_string, path), "File Permissions Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+
+    if (result == GUI::MessageBox::ExecYes) {
+        int access = 0;
+
+        if (requested_access & FileAccess::Read)
+            access |= O_RDONLY;
+
+        if (requested_access & FileAccess::Write)
+            access |= O_WRONLY;
+
+        if (requested_access & FileAccess::CreateOrRemove)
+            access |= O_CREAT;
+
+        if (requested_access & FileAccess::Truncate)
+            access |= O_TRUNC;
+
+        auto fd = open(path.characters(), access);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+
+            // TODO: check for root? or 1000
+
+            return { errno, Optional<IPC::File> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    if (result == GUI::MessageBox::ExecNo) {
+        // TODO: If a malicious program requests access over and over, do we eventually want to deny it?
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::PromptOpenFileResponse ClientConnection::prompt_open_file(String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+
+    Optional<String> user_picked_file = GUI::FilePicker::get_open_filepath(main_window, "Select file", path_to_view);
+
+    if (user_picked_file.has_value()) {
+        int access = 0;
+
+        if (requested_access & FileAccess::Read)
+            access |= O_RDONLY;
+
+        if (requested_access & FileAccess::Write)
+            access |= O_WRONLY;
+
+        if (requested_access & FileAccess::CreateOrRemove)
+            access |= O_CREAT;
+
+        auto fd = open(user_picked_file.value().characters(), access, 0644);
```

Comment:
As I already explained above, you could use `user_picked_file->characters()` here (not required of course).